src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,417 @@
+        {
+            BoundsCheck(BufferEnd, start);
+
+            var end = Seek(start, BufferEnd, length, false);
+
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at <paramref name="end"/> (inclusive).
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        /// <param name="end">The ending (inclusive) <see cref="SequencePosition"/> of the slice</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start, SequencePosition end)
+        {
+            BoundsCheck(BufferEnd, end);
+            BoundsCheck(end, start);
+
+            return SliceImpl(start, end);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="start"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="start">The starting (inclusive) <see cref="SequencePosition"/> at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(SequencePosition start)
+        {
+            BoundsCheck(BufferEnd, start);
+
+            return SliceImpl(start, BufferEnd);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="offset"/>, ending at the existing <see cref="ReadOnlyBuffer{T}"/>'s end.
+        /// </summary>
+        /// <param name="offset">The start index at which to begin this slice.</param>
+        public ReadOnlyBuffer<T> Slice(long offset)
+        {
+            if (offset == 0) return this;
+
+            var begin = Seek(BufferStart, BufferEnd, offset, false);
+            return SliceImpl(begin, BufferEnd);
+        }
+
+        /// <summary>
+        /// Copy the <see cref="ReadOnlyBuffer{T}"/> to the specified <see cref="Span{Byte}"/>.
+        /// </summary>
+        /// <param name="destination">The destination <see cref="Span{Byte}"/>.</param>
+        public void CopyTo(Span<T> destination)
+        {
+            if (Length > destination.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.destination);
+            }
+
+            foreach (var buffer in this)
+            {
+                buffer.Span.CopyTo(destination);
+                destination = destination.Slice(buffer.Length);
+            }
+        }
+
+        /// <summary>
+        /// Converts the <see cref="ReadOnlyBuffer{T}"/> to an array/>
+        /// </summary>
+        public T[] ToArray()
+        {
+            var buffer = new T[Length];
+            CopyTo(buffer);
+            return buffer;
+        }
+
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(byte))
```

Comment:
Can/Should we do similar for chars? but return text.

Reply:
Not even sure how valid is displaying chars for ROB<byte>, it helped us in Kestrel but for pure binary data might be better to use hex

Reply:
I think it would be good if the behavior of ToString on Span, Memory, and Buffer was the same, i.e. returns string if T is char.
@ahsonkhan, @ianhays, maybe there should be a shared routine that does the formatting?

Reply:
https://github.com/dotnet/corefx/blob/efdbbe6ba587a9efb1389f6d090b564f441c7b86/src/System.Memory/src/System/Span.cs#L334-L336

Reply:
We are in the process of changing ToString() to return the actual string values, for T = char.

See: https://github.com/dotnet/corefx/issues/26584#issuecomment-360615501

Reply:
What's the timeline for that work?

Reply:
I am working on it right now.

Reply:
We can merge this change with simplified `ToString()` and follow up when you change goes in.

Reply:
sounds good to me. just add a comment and file an issue.